src/full/Agda/Syntax/Abstract/UsedNames.hs

```diff
@@ -0,0 +1,182 @@
+
+module Agda.Syntax.Abstract.UsedNames
+  ( allUsedNames
+  ) where
+
+import Data.Foldable (foldMap)
+import Data.Semigroup (Semigroup, (<>))
+import Data.Set (Set)
+import qualified Data.Set as Set
+
+import Agda.Syntax.Common
+import Agda.Syntax.Abstract
+import Agda.Syntax.Concrete (FieldAssignment'(..))
+import Agda.Utils.List1 (List1)
+
+import Agda.Utils.Impossible
+
+-- | All names used in an abstract expression. This is used when rendering clauses to figure out
+--   which (implicit) pattern variables must be preserved. For example, the for @f : Nat → Nat@, the
+--   clause @f {n} = 0@ can be printed as @f = 0@ (dropping the @n@), but @f {n} = n@ must preserve
+--   the @n@.
+allUsedNames :: Expr -> Set Name
+allUsedNames = usedNames . boundAndUsed
+
+data BoundAndUsedNames = BoundAndUsedNames
+  { boundNames :: Set Name
+  , usedNames  :: Set Name }
```

Comment:
So my first reaction was: why commit to `Set` here?  
I have generalized similar name/variable gathering folds to `Monoid` or `Collection`-like structures, see`FreeVars`, `NamesIn`.
Then, I saw that we are not dealing with a monoid here, we have two monoidal operations: the `instance Semigroup BoundAndUsedNames` and the `parB`.
Here is a mathematical structure struggling to get out!
What are its laws?

I would very much like to see this traversal generalized to a suitable structure of collected results.  That's the way to increase confidence in your implementation!

Reply:
It feels like there ought to be some literature on binding structures that would be relevant.